src/Microsoft.OData.Client/ALinq/GroupByAnalyzer.cs

```diff
@@ -0,0 +1,476 @@
+            Debug.Assert(input != null, $"{nameof(input)} != null");
+            Debug.Assert(keySelector != null, $"{nameof(keySelector)} != null");
+
+            EnsureApplyInitialized(input);
+            Debug.Assert(input.Apply != null, $"{nameof(input)}.Apply != null");
+
+            // Scenario 1: GroupBy(d1 => [Constant])
+            ConstantExpression constExpr = keySelector.Body as ConstantExpression;
+            if (constExpr != null)
+            {
+                input.Apply.KeySelectorMap.Add(constExpr.Value.ToString(), constExpr);
+
+                return true;
+            }
+
+            // Scenario 2: GroupBy(d1 => d1.Property) and GroupBy(d1 = d1.NavProperty...Property)
+            if (keySelector.Body.NodeType == ExpressionType.MemberAccess)
+            {
+                MemberExpression memberExpr = (MemberExpression)keySelector.Body;
+
+                // Validate grouping expression
+                ResourceBinder.ValidationRules.ValidateGroupingExpression(memberExpr);
+
+                Expression boundExpression;
+                if (!TryBindToInput(input, keySelector, out boundExpression))
+                {
+                    return false;
+                }
+
+                input.Apply.GroupingExpressions.Add(boundExpression);
+                input.Apply.KeySelectorMap.Add(memberExpr.Member.Name, memberExpr);
+
+                return true;
+            }
+
+            // Only proceed beyond here if key selector body is a simple instantiation.
+            if (keySelector.Body.NodeType != ExpressionType.MemberInit && keySelector.Body.NodeType != ExpressionType.New)
+            {
+                return false;
+            }
+
+            // Scenario 2: GroupBy(d1 => new { d1.Property1, ..., d1.PropertyN })
+            // Scenario 3: GroupBy(d1 => new Cls { ClsProperty1 = d1.Property1, ..., ClsPropertyN = d1.PropertyN }) - not common but possible
+            GroupByKeySelectorAnalyzer.Analyze(input, keySelector);
+
+            return true;
+        }
+
+        /// <summary>
+        /// Ensure apply query option for the resource set is initialized
+        /// </summary>
+        /// <param name="input">The resource expression</param>
+        private static void EnsureApplyInitialized(QueryableResourceExpression input)
+        {
+            Debug.Assert(input != null, $"{nameof(input)} != null");
+
+            if (input.Apply == null)
+            {
+                ResourceBinder.AddSequenceQueryOption(input, new ApplyQueryOptionExpression(input.Type));
+            }
+        }
+
+        private static bool TryBindToInput(ResourceExpression input, LambdaExpression le, out Expression bound)
+        {
+            List<ResourceExpression> referencedInputs = new List<ResourceExpression>();
+            bound = InputBinder.Bind(le.Body, input, le.Parameters[0], referencedInputs);
+            if (referencedInputs.Count > 1 || (referencedInputs.Count == 1 && referencedInputs[0] != input))

```

Comment:
@habbes Because of how complex the expressions we're analyzing can be (e.g. the resource set can enclosed in an anonymously-typed transparent scope produced by a `SelectMany` operation), we call this method to confirm that the grouping expression (e.g. `d1.Product.Category.Name`) can be bound to the resource set expression (e.g. `ResourceSetExpression{"Sales"}`)

Reply:
@habbes Exactly